File: Data Structures & Algorithms/generate-parentheses/submission-11.java
class Solution {
    boolean isValid(String s) {
        int numOpen = 0;

        for (char c : s.toCharArray()) {
            if (c == '(') {
                numOpen++;
            } else {
                numOpen--;
            }

            if (numOpen < 0) {
                return false;
            }
        }
        
        return numOpen == 0;
    }

    void dfs(String s, List<String> res, int n) {
        if (s.length() == 2 * n) {
            if (isValid(s)) {
                res.add(s);
            }
            return;
        }

        dfs(s + '(', res, n);
        dfs(s + ')', res, n);
    }

    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        dfs("", res, n);
        return res;
    }
}

File: Data Structures & Algorithms/decode-ways/submission-11.py
class Solution:
    def numDecodings(self, s: str) -> int:
        # 1 2 1
        def dfs(i: int) -> int:
            if i == len(s):
                return 1
            if s[i] == '0':
                return 0

            # choose 1 character
            res = dfs(i + 1)

            if i < len(s) - 1:
                if s[i] == '1' or s[i] == '2' and s[i + 1] < '7':
                    # choose 2 characters
                    res += dfs(i + 2)

            return res
        
        return dfs(0)
